### ISSUE TITLE
MIDI sound support.

### ISSUE THREAD

**Issue by khokh2001** [chocolate-doom/chocolate-doom#639]
DMX library supports midi for sound playback(internally named GSS, likely refers Roland GS sound).

Example wad (DSPISTOL, DSWPNUP): https://drive.google.com/file/d/0B4n8cpbZ9qicU2kxRlRuYmQxZDA/view?usp=sharing

Roland demo:
https://www.youtube.com/watch?v=N-m0Is3kxK8

OPL3 demo:
https://www.youtube.com/watch?v=UKNCBazvOe4

There are 2 formats for midi sound:

Format 1 (WPNUP)  (similiar to pc speaker):
    +0(unsigned LE 16-bit int) format, always equals to 1
    +2(unsigned LE 16-bit int) sample length
    +4(unsigned LE 16-bit int) midi bank
    +6(unsigned LE 16-bit int) midi patch
    +8(array of bytes) samples
    
    1)sample rate is 140hz.
    2)sample controls the note key and pitch
    3)sample to note table:
  
   value   note   pitchbend
   0         note off
   1         29       127
   2         29       159
   3         30       127
   4         30       159
   5         31       127
   ....
   197     127     127
   198     127     159

Format 2 (PISTOL):
    Simply plays one note.
    
    +0(unsigned LE 16-bit int) format, always equals to 2
    +2(unsigned LE 16-bit int) midi bank
    +4(unsigned LE 16-bit int) midi patch
    +6(unsigned LE 16-bit int) note
    +8(unsigned LE 16-bit int) length

**Comment by fragglet** [comment]
Woah, interesting discovery. I never heard of this before. I assume you discovered this from disassembling the DMX code.

This would be nice to support in theory, but I suspect it's going to be difficult to support without completely restructuring much of the Chocolate Doom sound code.

**Comment by jmickle66666666** [comment]
Would it not be enough to "preprocess" the DMX midi samples and temporarily record all the sounds to a WAV or SND format and inject into the wad/memory?

**Comment by fragglet** [comment]
Point is that the MIDI samples sound different depending on your MIDI output (OPL3 vs. Roland, etc.). There's no real way to do what you suggest within the limits of the SDL_mixer MIDI API.

**Comment by linguica** [comment]
khokh2001, could you give a little more info on how this is used? I notice that loading either of your example WADs in doom2.exe in DOSBox does not work (all the sounds are still the normal digitized ones).

**Comment by khokh2001** [comment]
Set snd_sfxdevice to 6 for GM, or set to 2 for OPL.

**Comment by khokh2001** [comment]
When i googled GSS_init(from tic.map) i found this http://unix.superglobalmegacorp.com/source/dmx/

**Comment by jmickle66666666** [comment]
holy crap

**Comment by khokh2001** [comment]
GSS actually is "General Sound Set"

**Comment by fragglet** [comment]
Wow. That looks like it's the DMX source, right there.

**Comment by fragglet** [comment]
One brief message of caution I'd like to make: there is no license attached to this code, so it is not open source as defined under the [Open Source Definition](http://opensource.org/osd). As tempting as it may be, none of this code can be incorporated into Chocolate Doom unless we can somehow find the copyright owner and get it relicensed.

**Comment by khokh2001** [comment]
Paul Radek is owner of dmx: http://cocatalog.loc.gov/cgi-bin/Pwebrecon.cgi?Search_Arg=Dmx+sound+system&Search_Code=TALL&PID=pIh6244TaJUtUVnqyfeYPdWxbH7U&SEQ=20151111111850&CNT=25&HIST=1

**Comment by khokh2001** [comment]
I found his profile in LinkedIn https://www.linkedin.com/in/paulradek

**Comment by fragglet** [comment]
Thanks for the links. I think I found his Youtube/G+ account, so I've made an attempt to contact him. I will let you know if I get any response.

**Comment by AXDOOMER** [comment]
dmx34a is the DMX version used in Doom 1.666. 
dmx37lib would be the latest version used in Doom v1.7a, this is the last Doom version with sound code changes. The full source code of dmx37 is not included.

**Comment by khokh2001** [comment]
I think it'll not be too hard to reconstruct last version by side to side comparing 1.666 and 1.9 executables

**Comment by AXDOOMER** [comment]
I don't think there is a lot of changes anyway. Nice discovery, this is really going to help improving Chocolate-Doom's sound code. At least we can look at it.

**Comment by fragglet** [comment]
I kind of want to see if we can compile the Heretic/Hexen sources. They have all the hooks for DMX, and the Watcom compilers are open source nowadays.

**Comment by khokh2001** [comment]
It's possible. For asm code you should use tasm. I've already compiled heretic with Apogee Sound System throw simple dmx wrapper.

**Comment by fragglet** [comment]
Sounds impressive. Have you shared your wrapper at all?

**Comment by khokh2001** [comment]
I'll publish it tomorrow. Also i have ported linuxdoom to dos with original video system and with dmx wrapper sound. I'll try to publish it too.

**Comment by fragglet** [comment]
That sounds really cool.

**Comment by fragglet** [comment]
By "original video system", do you mean the page flipping code that Lee Killough describes here:

https://github.com/fragglet/mbf/blob/master/i_video.c#L488

AFAIK it's not in the Heretic/Hexen source.

**Comment by khokh2001** [comment]
Yes. I've reverse engineered it. But i implemented column and span drawing function in C.  While vanilla doom has assembler versions. (planar.asm)

**Comment by khokh2001** [comment]
Just compiled heretic with dmx. It really works. Even random pitch works.

**Comment by khokh2001** [comment]
pcdoom port: https://github.com/khokh2001/pcdoom

**Comment by maraakate** [comment]
I'm the one who originally got this code from Romero and I have been maintaing DOS ports for a few years (no public releases) at:

https://bitbucket.org/maraakate/heretic-for-dos
https://bitbucket.org/maraakate/doomnew-for-dos

I was kind of wondering how long it would take until someone would accidentally find it.  I purposely posted it on some telnet BBS' hoping a fellow doom fan would find it there first.

**Comment by jmtd** [comment]
The DMX source URL is now behind HTTP auth.

**Comment by AXDOOMER** [comment]
That's my fault. I asked the site's author about where he found the sound code. 

![doom](https://cloud.githubusercontent.com/assets/6194072/11119717/7c958e6c-8919-11e5-9d81-12aee5069afb.png)

I've downloaded a copy of DMX.zip and I guess many of us made a copy.

**Comment by haleyjd** [comment]
@khokh2001 Your work in I_IBM.C matches the disassembly from Strife almost perfectly; really good work. Anybody interested in how the non-linear Mode Y blit and dirty rects refresh system worked should look at this.

**Comment by maraakate** [comment]
I'm the one who originally got this code from Romero and I have been 
maintaing DOS ports for a few years (no public releases) at:

https://bitbucket.org/maraakate/heretic-for-dos
https://bitbucket.org/maraakate/doomnew-for-dos

I was kind of wondering how long it would take until someone would 
accidentally find it. I purposely posted it on some telnet BBS' hoping a 
fellow doom fan would find it there first.

Romero originally gave me the libs + source as a gift for my work on 
Daikatana v1.3.

My Doom and Heretic Watcom updates aren't extreme: some Coop updates, 
some unfinished, but mostly working DEH code from chocolate doom, 
SimulEyes VR 3D goggles support by using the library from the released 
Greed source code.

The doom update is also cross-compilable with DJGPP, but currently DMX 
is not finished being ported to DJGPP.  Neozeed and I were working to 
update it to DJGPP and then make a public release, but it appears the 
URL got snatched by a search engine.

The doom code also uses I_IBM.C from Heretic and modified to match 
changes, but no actual disassembly has been done.  The code works 
properly as the original v1.9 as far as I can tell, and I have tested it 
on Kali with friends over real DOS computers with no ill-effects.

Maraakate

**Comment by maraakate** [comment]
I should also mention that attempts to contact Paul Radek have failed.  My port is technically "illegal" at this point, but if DMX can be updated to DJGPP I had plans to make it work as a DXE (DJGPP's version of a DLL) to circumvent the issue, but finding the DXE would be up to the user.

**Comment by fragglet** [comment]
When you say you're "the one who originally got this code from Romero" which code are you referring to - the DMX sources?

**Comment by maraakate** [comment]
Yes, the DMX source and libraries.  He originally got it from someone at Raven (since his original copy appeared to be lost).  So that may explain the older version.  I am unsure, but Heretic and Hexen may have been built off a slightly older version of Doom source code.

**Comment by fragglet** [comment]
It's great to have this source regardless of what the legality may be. I just wish we could contact Paul Radek and get his permission to GPL it - then it could be a lot more useful.

Side note: your port is technically violating the GPL as you're combining GPLed code from Chocolate Doom with the non-free DMX code. So I grant you the permission to use Chocolate Doom code in your port provided that you're releasing the source. I don't have any problem with people reusing my code as long as they're sharing alike and contributing back, like I do with Chocolate Doom.

**Comment by maraakate** [comment]
The source is publicly available via the linked github repositories.  I never really intended to release this until the DMX code was finished being ported to DJGPP and if Paul Radek could ever be contacted.

For hackers: you can change the Doom 1 code to use 22khz instead of 11khz for SB by default.  There is a define in one of the headers for Doom 1 and Heretic itself.  GUS uses the weirdo 19khz as it's minimum frequency, so upping it to 22khz doesn't do a whole lot for it; but it is nice for our SB friends to finally hear the sound effects on comparable quality to the GUS.  There is also another setenv for some verbose debugging output from the DMX library.  I forget specifically what it is but doing a grep for setenv should probably find it for you.

Regarding Chocolate Doom: I have nothing special to contribute back.  I only really used the DEH code and some of the chex quest updates iirc and I have not found bugs in their implementations.  I worked basically from scratch... Taking linuxdoom and then merging part of heretic's i_ibm and friends to make it all work.  At the time, I was still relatively new to coding so it took me almost a year of off and on work to actually make it work properly.  I also noticed there was lots of little bugs in that linux guys updates to the 1.9 code before public release but I assume all of those bugs have been fixed long ago.

**Comment by nukeykt** [comment]
Did you found any trace of Paul Radek?

**Comment by fragglet** [comment]
Nope. Heard nothing.

**Comment by sneakernets** [comment]
I can contact him through LinkedIn if that is a desirable route.

**Comment by fragglet** [comment]
We've made multiple attempts at contact and I'm concerned that continuing efforts may seem like harassment.

**Comment by nukeykt** [comment]
Found out, that "Raptor: Call of the Shadows" had GSS sounds: https://youtu.be/iTwtIozXjQo

**Comment by jherrera80** [comment]
I managed to compile heretic with dmx (with OpenWatcom 1.9). Music and sfx are working OK. Problem is that TSM_NewService call always results in "Stack Overflow", so I have to comment that, comment the I_StopSong kludge (cause ticcount is not incremented anymore by dmx) and write my own I_GetTime():

	ftime(&now);
	ticcount = ((now.time * 1000 + now.millitm) * 35) / 1000;

I tried with all versions of dmx and all of them give me a stack overflow when calling TSM_NewService. Is there any way to solve this?

Thinking about it, it's kinda silly to have dmx control the 35hz timer, me thinks.

**Comment by maraakate** [comment]
jherrera80: email me kprxradio@gmail.com for further details.

**Comment by datajake1999** [comment]
I know this is an old thread, but I got in contact with Paul Radek via LinkedIn. He told me he isn't opposed to open sourcing the DMX code, he just needs to find the time to clean out some stupid comments. He said he could use some help, and I told him that I am willing to help him with this as the DMX code would greatly benefit the Doom porting community.

**Comment by sneakernets** [comment]
> know this is an old thread, but I got in contact with Paul Radek via LinkedIn. He told me he isn't opposed to open sourcing the DMX code, he just needs to find the time to clean out some stupid comments. He said he could use some help, and I told him that I am willing to help him with this as the DMX code would greatly benefit the Doom porting community.

This is wonderful news.

**Comment by OpenRift412** [comment]
> I know this is an old thread, but I got in contact with Paul Radek via LinkedIn. He told me he isn't opposed to open sourcing the DMX code, he just needs to find the time to clean out some stupid comments. He said he could use some help, and I told him that I am willing to help him with this as the DMX code would greatly benefit the Doom porting community.

Did anything ever become of this?

**Comment by datajake1999** [comment]
I haven't gotten an update from Paul as of late. I was clearly expecting this since he has more important things on his plate.

**Comment by OpenRift412** [comment]
> I haven't gotten an update from Paul as of late. I was clearly expecting this since he has more important things on his plate.

Ah. Yeah, had a similar thing when trying to get in contact with Kee Kimbrell about DWANGO stuff. Though that was like a year ago, don't know if he forgot or straight up ghosted me...
